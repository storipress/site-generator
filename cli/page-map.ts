import { Array, Effect, Option, Record, Tuple, pipe } from 'effect'
import { renderPage } from '@storipress/render-helper/index.mts'
import { RawOtherPageDataSchema } from '../schema/other-page.ts'
import type { OtherPage } from '../schema/other-page.ts'
import { BlockMapping } from './block-mapping.ts'
import { FileSystem } from './services/FileSystem.ts'
import type { GetPagesQuery } from './handler/graphql-operations.ts'
import type { MappingResult } from './types.ts'
import { generatedPath, parseSchemaJson } from './helpers.ts'
import { PAGE_FALLBACK, createTemplateInfo } from './utils/MappingInfo.ts'

export function createPagesMapping(pageList: GetPagesQuery['pages'], mappingResult: MappingResult) {
  return Effect.runPromise(pipe(createPagesMapping$(pageList, mappingResult), Effect.provide(FileSystem.Live)))
}

export function createPagesMapping$(
  pageList: GetPagesQuery['pages'],
  mappingResult: MappingResult,
): Effect.Effect<Record<string, OtherPage>, never, FileSystem> {
  if (pageList.length === 0) {
    return Effect.succeed({})
  }

  return pipe(
    pageList,
    Array.map((page) => {
      return pipe(
        Effect.gen(function* ($) {
          const fs = yield* $(FileSystem)
          const current = yield* $(parseSchemaJson(RawOtherPageDataSchema, page.current))
          const seo = JSON.parse(page.seo)
          const title = page.title
          const { html, segments } = current ? renderPage(current.editorContent) : { html: undefined, segments: [] }
          const layout = {
            id: page.id,
            template: current?.template,
          }

          const { info, templateDir: pageDir } = createTemplateInfo({
            id: page.id,
            template: current?.template,
            fallback: PAGE_FALLBACK,
            path: `./article/other-template-${page.id}`,
            pathMapping: BlockMapping.page,
          })
          const destDir = generatedPath`article/other-template-${page.id}`

          mappingResult.page.push(info)

          const pageExist = yield* $(fs.exists(pageDir))

          if (!pageExist) {
            return Option.none()
          }

          yield* $(fs.mkdir(destDir, { recursive: true }), Effect.unlessEffect(fs.exists(destDir)))

          yield* $(fs.copy(pageDir, destDir))

          return Option.some(
            Tuple.make(seo.slug as string, {
              title,
              seo,
              pageTitle: current?.templateData?.title,
              cover: {
                url: current?.templateData?.headlineURL,
                alt: current?.templateData?.headlineAlt,
                caption: current?.templateData?.caption,
              },
              layout,
              html: html ?? '',
              segments,
            } satisfies OtherPage),
          )
        }),
        Effect.annotateLogs('pageId', page.id),
      )
    }),
    Effect.all,
    Effect.map((optionTuples) => pipe(optionTuples, Array.getSomes, Record.fromEntries)),
  )
}
