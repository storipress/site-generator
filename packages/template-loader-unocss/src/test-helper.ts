import uno from '@unocss/preset-uno'
import { h } from '@unocss/preset-mini/utils'
import unocss from 'unocss/vite'
import { build } from 'vite'
import type { OutputAsset, RollupOutput } from 'rollup'
import transformerDirectives from '@unocss/transformer-directives'
import vue from '@vitejs/plugin-vue'
import type { NestedHooks } from 'hookable'
import transformer from './transform'
import type { TransformerHooks } from './transform'

interface CreateTransformBuildInput {
  code: string
  pathPrefix?: string
  hooks?: NestedHooks<TransformerHooks>
}

export async function createTransformBuild({ code, pathPrefix, hooks }: CreateTransformBuildInput) {
  let vueCode = ''
  const res = await build({
    logLevel: 'error',
    configFile: false,
    plugins: [
      {
        name: 'virtual',
        resolveId(id) {
          if (id.endsWith('entry')) {
            return id
          }
          if (id.endsWith('entry.vue')) {
            return id
          }
        },
        load(id) {
          if (id.endsWith('entry')) {
            return `
              export {default} from '${pathPrefix ?? ''}/entry.vue'
            `
          }
          if (id.endsWith('entry.vue')) {
            return code
          }
        },
      },
      unocss({
        mode: 'vue-scoped',
        presets: [
          uno({
            preflight: false,
          }),
        ],
        rules: [[/^ff-(.+)$/, ([, family]) => ({ 'font-family': `"${h.bracket(family)}"` })]],
        transformers: [
          transformer({
            hooks,
          }),
          transformerDirectives(),
        ],
      }),
      {
        name: 'inspect',
        transform(code, id) {
          if (id.endsWith('entry.vue')) {
            vueCode = code
          }
          return code
        },
      },
      vue({}),
    ],
    build: {
      lib: {
        entry: 'entry',
        formats: ['es'],
      },
      rollupOptions: {
        output: undefined,
        external: ['vue'],
      },
    },
  })

  const out = res as RollupOutput[]

  return {
    code: out[0].output[0].code,
    vue: vueCode,
    css: (out[0].output[1] as OutputAsset)?.source || '',
  }
}
