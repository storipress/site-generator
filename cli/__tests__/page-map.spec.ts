import { expect, it } from 'vitest'
import { Effect, pipe } from 'effect'
import { createPagesMapping$ } from '../page-map'
import type { RawOtherPageData } from '../../schema/other-page'
import type { MappingResult } from '../types'
import { createTestLayer } from './test-helper'

it('should prepare page template and without using fallback', async () => {
  const mapping: MappingResult = {
    article: [],
    front: [],
    page: [],
  }
  const { TestLayer, fileSystemActions } = createTestLayer()
  const res = await pipe(
    createPagesMapping$(
      [
        {
          id: '1',
          title: 'title',
          seo: JSON.stringify({ slug: 'page-slug' }),
          current: JSON.stringify({
            elements: {
              blockquote: 'blockquote',
              dropcap: 'dropcap',
            },
            template: 'spare-page-3',
            templateData: {
              title: 'pageTitle',
              description: 'description',
              headlineURL: 'headlineURL',
              headlineAlt: 'headlineAlt',
              caption: 'caption',
            },
            editorContent: {
              type: 'doc',
              content: [
                {
                  type: 'paragraph',
                  content: [
                    {
                      type: 'text',
                      text: 'text',
                    },
                  ],
                },
              ],
            },
          } satisfies RawOtherPageData),
        },
      ],
      mapping,
    ),
    Effect.provide(TestLayer),
    Effect.runPromise,
  )

  expect(res).toMatchSnapshot()
  expect(mapping.page[0].isFallback).toBe(false)
  expect(mapping).toMatchSnapshot()
  expect(fileSystemActions).toMatchSnapshot()
})

it('should prepare page template and use fallback', async () => {
  const mapping: MappingResult = {
    article: [],
    front: [],
    page: [],
  }
  const { TestLayer, fileSystemActions } = createTestLayer()
  const res = await pipe(
    createPagesMapping$(
      [
        {
          id: '1',
          title: 'title',
          seo: JSON.stringify({ slug: 'page-slug' }),
          current: JSON.stringify({
            elements: {
              blockquote: 'blockquote',
              dropcap: 'dropcap',
            },
            template: 'template',
            templateData: {
              title: 'pageTitle',
              description: 'description',
              headlineURL: 'headlineURL',
              headlineAlt: 'headlineAlt',
              caption: 'caption',
            },
            editorContent: {
              type: 'doc',
              content: [
                {
                  type: 'paragraph',
                  content: [
                    {
                      type: 'text',
                      text: 'text',
                    },
                  ],
                },
              ],
            },
          } satisfies RawOtherPageData),
        },
      ],
      mapping,
    ),
    Effect.provide(TestLayer),
    Effect.runPromise,
  )

  expect(res).toMatchSnapshot()
  expect(mapping.page[0].isFallback).toBe(true)
  expect(mapping).toMatchSnapshot()
  expect(fileSystemActions).toMatchSnapshot()
})

it('can handle optional templateData', async () => {
  const mapping: MappingResult = {
    article: [],
    front: [],
    page: [],
  }
  const { TestLayer, fileSystemActions } = createTestLayer()
  const res = await pipe(
    createPagesMapping$(
      [
        {
          id: '1',
          title: 'title',
          seo: JSON.stringify({ slug: 'page-slug' }),
          current: JSON.stringify({
            elements: {
              blockquote: 'blockquote',
              dropcap: 'dropcap',
            },
            template: 'spare-page-3',
            templateData: {
              title: 'pageTitle',
            },
            editorContent: {
              type: 'doc',
              content: [
                {
                  type: 'paragraph',
                  content: [
                    {
                      type: 'text',
                      text: 'text',
                    },
                  ],
                },
              ],
            },
          } satisfies RawOtherPageData),
        },
      ],
      mapping,
    ),
    Effect.provide(TestLayer),
    Effect.runPromise,
  )

  expect(res).toMatchSnapshot()
  expect(mapping.page[0].isFallback).toBe(false)
  expect(mapping).toMatchSnapshot()
  expect(fileSystemActions).toMatchSnapshot()
})

it('can handle empty list', async () => {
  const mapping: MappingResult = {
    article: [],
    front: [],
    page: [],
  }
  const { TestLayer, fileSystemActions } = createTestLayer()
  const res = await pipe(createPagesMapping$([], mapping), Effect.provide(TestLayer), Effect.runPromise)

  expect(res).toEqual({})
  expect(mapping).toEqual({
    article: [],
    front: [],
    page: [],
  })
  expect(fileSystemActions).toEqual([])
})

it('can handle empty page', async () => {
  const { TestLayer, fileSystemActions } = createTestLayer()
  const mapping: MappingResult = {
    article: [],
    front: [],
    page: [],
  }
  const res = await pipe(
    createPagesMapping$(
      [
        {
          id: '1',
          title: 'title',
          seo: JSON.stringify({ slug: 'page-slug' }),
          current: null,
        },
      ],
      mapping,
    ),
    Effect.provide(TestLayer),
    Effect.runPromise,
  )

  expect(res).toMatchInlineSnapshot(`
    {
      "page-slug": {
        "cover": {
          "alt": undefined,
          "caption": undefined,
          "url": undefined,
        },
        "html": "",
        "layout": {
          "id": "1",
          "template": undefined,
        },
        "pageTitle": undefined,
        "segments": [],
        "seo": {
          "slug": "page-slug",
        },
        "title": "title",
      },
    }
  `)
  expect(mapping).toMatchInlineSnapshot(`
    {
      "article": [],
      "front": [],
      "page": [
        {
          "actualTemplate": "other-page-nophoto",
          "id": "1",
          "isFallback": true,
          "path": "./article/other-template-1",
          "template": undefined,
        },
      ],
    }
  `)
  expect(fileSystemActions).toMatchInlineSnapshot(`
    [
      {
        "action": "exists",
        "path": "#unknown",
      },
      {
        "action": "mkdir",
        "options": {
          "recursive": true,
        },
        "path": "/article/other-template-1",
      },
      {
        "action": "exists",
        "path": "/article/other-template-1",
      },
      {
        "action": "copy",
        "from": "#unknown",
        "to": "/article/other-template-1",
      },
    ]
  `)
})
