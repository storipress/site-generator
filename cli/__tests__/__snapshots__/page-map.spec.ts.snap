// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`can handle optional templateData 1`] = `
{
  "page-slug": {
    "cover": {
      "alt": undefined,
      "caption": undefined,
      "url": undefined,
    },
    "html": "<p>text</p>",
    "layout": {
      "id": "1",
      "template": "spare-page-3",
    },
    "pageTitle": "pageTitle",
    "segments": [
      {
        "html": "<p>text</p>",
        "id": "normal",
        "type": "p",
      },
    ],
    "seo": {
      "slug": "page-slug",
    },
    "title": "title",
  },
}
`;

exports[`can handle optional templateData 2`] = `
{
  "article": [],
  "front": [],
  "page": [
    {
      "actualTemplate": "spare-page-3",
      "id": "1",
      "isFallback": false,
      "path": "./article/other-template-1",
      "template": "spare-page-3",
    },
  ],
}
`;

exports[`can handle optional templateData 3`] = `
[
  {
    "action": "exists",
    "path": "#unknown",
  },
  {
    "action": "mkdir",
    "options": {
      "recursive": true,
    },
    "path": "/article/other-template-1",
  },
  {
    "action": "exists",
    "path": "/article/other-template-1",
  },
  {
    "action": "copy",
    "from": "#unknown",
    "to": "/article/other-template-1",
  },
]
`;

exports[`should prepare page template and use fallback 1`] = `
{
  "page-slug": {
    "cover": {
      "alt": "headlineAlt",
      "caption": "caption",
      "url": "headlineURL",
    },
    "html": "<p>text</p>",
    "layout": {
      "id": "1",
      "template": "template",
    },
    "pageTitle": "pageTitle",
    "segments": [
      {
        "html": "<p>text</p>",
        "id": "normal",
        "type": "p",
      },
    ],
    "seo": {
      "slug": "page-slug",
    },
    "title": "title",
  },
}
`;

exports[`should prepare page template and use fallback 2`] = `
{
  "article": [],
  "front": [],
  "page": [
    {
      "actualTemplate": "other-page-nophoto",
      "id": "1",
      "isFallback": true,
      "path": "./article/other-template-1",
      "template": "template",
    },
  ],
}
`;

exports[`should prepare page template and use fallback 3`] = `
[
  {
    "action": "exists",
    "path": "#unknown",
  },
  {
    "action": "mkdir",
    "options": {
      "recursive": true,
    },
    "path": "/article/other-template-1",
  },
  {
    "action": "exists",
    "path": "/article/other-template-1",
  },
  {
    "action": "copy",
    "from": "#unknown",
    "to": "/article/other-template-1",
  },
]
`;

exports[`should prepare page template and without using fallback 1`] = `
{
  "page-slug": {
    "cover": {
      "alt": "headlineAlt",
      "caption": "caption",
      "url": "headlineURL",
    },
    "html": "<p>text</p>",
    "layout": {
      "id": "1",
      "template": "spare-page-3",
    },
    "pageTitle": "pageTitle",
    "segments": [
      {
        "html": "<p>text</p>",
        "id": "normal",
        "type": "p",
      },
    ],
    "seo": {
      "slug": "page-slug",
    },
    "title": "title",
  },
}
`;

exports[`should prepare page template and without using fallback 2`] = `
{
  "article": [],
  "front": [],
  "page": [
    {
      "actualTemplate": "spare-page-3",
      "id": "1",
      "isFallback": false,
      "path": "./article/other-template-1",
      "template": "spare-page-3",
    },
  ],
}
`;

exports[`should prepare page template and without using fallback 3`] = `
[
  {
    "action": "exists",
    "path": "#unknown",
  },
  {
    "action": "mkdir",
    "options": {
      "recursive": true,
    },
    "path": "/article/other-template-1",
  },
  {
    "action": "exists",
    "path": "/article/other-template-1",
  },
  {
    "action": "copy",
    "from": "#unknown",
    "to": "/article/other-template-1",
  },
]
`;
